README.md
# Light-weight image processor for NodeJS

[![npm versions](https://img.shields.io/npm/v/@randy.tarampi/lwip.svg?style=flat-square)](https://www.npmjs.org/package/@randy.tarampi/lwip)
[![npm downloads](https://img.shields.io/npm/dt/@randy.tarampi/lwip.svg?style=flat-square)](https://www.npmjs.com/package/@randy.tarampi/lwip)
[![npm license](https://img.shields.io/npm/l/@randy.tarampi/lwip.svg?registry_uri=https%3A%2F%2Fregistry.npmjs.com&style=flat-square)](https://www.npmjs.com/package/@randy.tarampi/lwip)
[![Build status](https://img.shields.io/travis/com/randytarampi/lwip.svg?style=flat-square)](https://travis-ci.com/randytarampi/lwip)
[![Build status](https://ci.appveyor.com/api/projects/status/46mk5218x995svhw/branch/master?svg=true&style=flat-square)](https://ci.appveyor.com/project/randytarampi/lwip/branch/master)
[![Coverage status](https://img.shields.io/coveralls/randytarampi/lwip.svg?style=flat-square)](https://coveralls.io/github/randytarampi/lwip?branch=master)
[![Maintainability status](https://img.shields.io/codeclimate/maintainability-percentage/randytarampi/lwip.svg?style=flat-square)](https://codeclimate.com/github/randytarampi/lwip/maintainability)
[![Analytics](https://ga-beacon.appspot.com/UA-50921068-1/beacon/github/randytarampi/lwip/?flat&useReferrer)](https://github.com/igrigorik/ga-beacon)
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Frandytarampi%2Flwip.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Frandytarampi%2Flwip?ref=badge_shield)

[![Join the chat at https://gitter.im/EyalAr/lwip](https://img.shields.io/gitter/room/EyalAr/lwip.svg?style=flat-square)](https://gitter.im/EyalAr/lwip?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[![Install @randy.tarampi/lwip](https://nodeico.herokuapp.com/@randy.tarampi/lwip.svg)](https://www.npmjs.com/package/@randy.tarampi/lwip)


This is a branch based off of [@kant2002/lwip](https://www.npmjs.com/package/@kant2002/lwip) and [@mcph/lwip](https://www.npmjs.com/package/@mcph/lwip), customized for my own use.

0. [Overview](#overview)
  0. [Installation](#installation)
  0. [Usage](#usage)
  0. [Supported formats](#supported-formats)
  0. [Colors specification](#colors-specification)
  0. [Note on transparent images](#note-on-transparent-images)
  0. [Note on threading performance](#note-on-threading-performance)
0. [API](#api)
  0. [Open an image from file or buffer](#open-an-image)
  0. [Create a new blank image](#create-a-new-image)
  0. [Image operations](#image-operations)
    0. [Resize](#resize)
    0. [Scale](#scale)
    0. [Contain](#contain)
    0. [Cover](#cover)
    0. [Rotate](#rotate)
    0. [Crop](#crop)
    0. [Blur](#blur)
    0. [Sharpen](#sharpen)
    0. [Mirror](#mirror)
    0. [Flip](#flip)
    0. [Border](#border)
    0. [Pad](#pad)
    0. [Adjust saturation](#saturate)
    0. Adjust lightness: [lighten](#lighten) / [darken](#darken)
    0. [Adjust hue](#hue)
    0. [Fade (adjust transparency)](#fade)
    0. [Opacify](#opacify)
    0. [Paste](#paste)
    0. [Set pixel](#set-pixel)
    0. [Set metadata](#set-metadata)
  0. [Getters](#getters)
    0. [Width](#width)
    0. [Height](#height)
    0. [Pixel](#get-pixel)
    0. [Clone](#clone)
    0. [Extract / Copy](#extract)
    0. [Get as a Buffer](#get-as-a-buffer)
      0. [JPEG](#jpeg)
      0. [PNG](#png)
      0. [GIF](#gif)
    0. [Write to file](#write-to-file)
    0. [Get metadata](#get-metadata)
    0. [Get dominant color](#get-dominant-color)
  0. [Batch operations](#batch-operations)
0. [Copyrights](#copyrights)

## Overview

This module provides comprehensive, fast, and simple image processing and
manipulation capabilities.

**There are no external runtime dependencies**, which means you don't have to
install anything else on your system.

**This module is in active development. [New features](https://github.com/EyalAr/lwip/issues?labels=feature+request&page=1&state=open) are being added.**

[Read the background for the development of this module.](http://eyalarubas.com/image-processing-nodejs.html)

### Installation

```bash
npm install @randy.tarampi/lwip
```

Or, clone this repo and `cd @randy.tarampi/lwip && npm install`.

You can run tests with `npm test`.

##### NOTE: Installation of this module involves compiling native code
If `npm install @randy.tarampi/lwip` fails, you probably need to setup your system.

See the [instructions for `node-gyp`](https://github.com/TooTallNate/node-gyp#installation).

Building on Windows with Visual Studio requires version 2013 or higher.

### Usage

**Typical workflow:**

0. Open an image and get an image object.
0. Manipulate it.
0. Save to disk / Send image buffer over network / etc.

**Example (batch operations):**

```Javascript
// obtain an image object:
require('@randy.tarampi/lwip').open('image.jpg', function(err, image){

  // check err...
  // define a batch of manipulations and save to disk as JPEG:
  image.batch()
    .scale(0.75)          // scale to 75%
    .rotate(45, 'white')  // rotate 45degs clockwise (white fill)
    .crop(200, 200)       // crop a 200X200 square from center
    .blur(5)              // Gaussian blur with SD=5
    .writeFile('output.jpg', function(err){
      // check err...
      // done.
    });

});
```

**Example (non-batch):**

```Javascript
var lwip = require('@randy.tarampi/lwip');

// obtain an image object:
lwip.open('image.jpg', function(err, image){

  // check err...
  // manipulate image:
  image.scale(0.5, function(err, image){

    // check err...
    // manipulate some more:
    image.rotate(45, 'white', function(err, image){

      // check err...
      // encode to jpeg and get a buffer object:
      image.toBuffer('jpg', function(err, buffer){

        // check err...
        // save buffer to disk / send over network / etc.

      });

    });

  });

});
```

### Supported formats

**Decoding (reading):**

- JPEG, 1 & 3 channels (grayscale & RGB).
- PNG, transparency supported.
- GIF, transparency supported. Animated GIFs can be read, but only the first
  frame will be retrieved.

**Encoding (writing):**

- JPEG, 3 channels (RGB).
- PNG (lossless), 3 channels (RGB) or 4 channels (RGBA).
- GIF (no animations)

Other formats may also be supported in the future, but are probably less urgent.
Check the issues to see [which formats are planned to be supported](https://github.com/EyalAr/lwip/issues?labels=format+request&page=1&state=open).
Open an issue if you need support for a format which is not already listed.

### Colors specification

In LWIP colors are coded as RGBA values (red, green, blue and an alpha channel).

Colors are specified in one of three ways:

- As a string. possible values:

  ```Javascript
  "black"    // {r: 0, g: 0, b: 0, a: 100}
  "white"    // {r: 255, g: 255, b: 255, a: 100}
  "gray"     // {r: 128, g: 128, b: 128, a: 100}
  "red"      // {r: 255, g: 0, b: 0, a: 100}
  "green"    // {r: 0, g: 255, b: 0, a: 100}
  "blue"     // {r: 0, g: 0, b: 255, a: 100}
  "yellow"   // {r: 255, g: 255, b: 0, a: 100}
  "cyan"     // {r: 0, g: 255, b: 255, a: 100}
  "magenta"  // {r: 255, g: 0, b: 255, a: 100}
  ```

- As an array `[R, G, B, A]` where `R`, `G` and `B` are integers between 0 and
  255 and `A` is an integer between 0 and 100.
- As an object `{r: R, g: G, b: B, a: A}` where `R`, `G` and `B` are integers
  between 0 and 255 and `A` is an integer between 0 and 100.

**Note**: The `A` value (alpha channel) is always optional and defaults to
100 (completely opaque).

### Note on transparent images

0. Transparency is supported through an alpha channel which ranges between 0
   and 100. 0 is completely transparent and 100 is completely opaque.
0. Not all formats support transparency. If an image with an alpha channel is
   encoded with a format which does not support transparency, the alpha channel
   will be ignored (effectively setting it to 100% for all pixels).

### Note on threading performance

All operations are asynchronous, and processing takes place in a thread pool
managed by libuv which is part of NodeJS. This thread pool is separate from the event
loop used to process HTTP requests, so use of lwip should not significantly affect the
handling of HTTP requests by a web application. The thread pool is however shared with
other threaded native modules such as those providing database and filesystem IO.

The default thread pool size of 4 will be appropriate for most applications. However
if your application regularly processes many images concurrently and and you wish
to take full advantage of a multicore system or prevent heavy image processing work
from delaying database or filesystem IO, you may want to increase the size of the
thread pool by setting the UV_THREADPOOL_SIZE environmental variable to the NodeJS
process, e.g.:

```
UV_THREADPOOL_SIZE=8 node your_script.js
```

## API

All operations are done on an `image` object. An `image` object can be obtained
by:

0. Openning an existing image file or buffer with the [`open`](#open-an-image)
   method.
0. Creating a new image object with the [`create`](#create-a-new-image) method.
0. Cloning an existing image object with the [`image.clone`](#clone) method.
0. Extracting a sub-image from an existing image object with the
   [`image.extract`](#extract) method.

### Open an image

`lwip.open(source, type, callback)`

0. `source {String/Buffer}`: The path to the image on disk or an image buffer.
0. `type {String/Object}`: **Optional** type of the image. If omitted, the type
   will be inferred from the file extension. If `source` is a buffer, `type`
   must be specified. If `source` is an encoded image buffer, `type` must be
   a string of the image type (i.e. `"jpg"`). If `source` is a raw pixels buffer
   `type` must be an object with `type.width` and `type.height` properties.
0. `callback {Function(err, image)}`

**Note about raw pixels buffers:** `source` may be a buffer of raw pixels. The
buffer may contain pixels of 1-4 channels, where:

0. 1 channel is a grayscale image.
0. 2 channels is a grayscale image with an alpha channel.
0. 3 channels is an RGB image.
0. 4 channels is an RGBA image (with an alpha channel).

In other words, if the image in the buffer has width `W` and height `H`, the
size of the buffer can be `W*H`, `2*W*H`, `3*W*H` or `4*W*H`.

The channel values in the buffer must be stored sequentially. I.e. first all the
Red values, then all the Green values, etc.

#### Open file example

```Javascript
const lwip = require('@randy.tarampi/lwip');
lwip.open('path/to/image.jpg', function(err, image){
    // check 'err'. use 'image'.
    // image.resize(...), etc.
});
```

#### Open buffer example

```Javascript
const fs = require('fs'),
    lwip = require('@randy.tarampi/lwip');

fs.readFile('path/to/image.png', function(err, buffer){
  // check err
  lwip.open(buffer, 'png', function(err, image){
      // check 'err'. use 'image'.
      // image.resize(...), etc.
  });
});
```

### Create a new image

`lwip.create(width, height, color, callback)`

0. `width {Integer>0}`: The width of the new image.
0. `height {Integer>0}`: The height of the new image.
0. `color {String / Array / Object}`: **Optional** Color of the canvas. See
   [colors specification](#colors-specification). Defaults to a transparent
   canvas `{r:0, g:0, b:0, a:0}`.
0. `callback {Function(err, image)}`

**Example**:

```Javascript
const lwip = require('@randy.tarampi/lwip');

lwip.create(500, 500, 'yellow', function(err, image){
  // check err
  // 'image' is a 500X500 solid yellow canvas.
});
```

### Image operations

#### Resize

`image.resize(width, height, inter, callback)`

0. `width {Integer}`: Width in pixels.
0. `height {Integer}`: **Optional** height in pixels. If omitted, `width` will
   be used.
0. `inter {String}`: **Optional** interpolation method. Defaults to `"lanczos"`.
   Possible values:
   - `"nearest-neighbor"`
   - `"moving-average"`
   - `"linear"`
   - `"grid"`
   - `"cubic"`
   - `"lanczos"`
0. `callback {Function(err, image)}`

#### Scale

`image.scale(wRatio, hRatio, inter, callback)`

0. `wRatio {Float}`: Width scale ratio.
0. `hRatio {Float}`: **Optional** height scale ratio. If omitted, `wRatio` will
   be used.
0. `inter {String}`: **Optional** interpolation method. Defaults to `"lanczos"`.
   Possible values:
   - `"nearest-neighbor"`
   - `"moving-average"`
   - `"linear"`
   - `"grid"`
   - `"cubic"`
   - `"lanczos"`
0. `callback {Function(err, image)}`

#### Contain

Contain the image in a colored canvas. The image will be resized to the largest
possible size such that it's fully contained inside the canvas.

`image.contain(width, height, color, inter, callback)`

0. `width {Integer}`: Canvas' width in pixels.
0. `height {Integer}`: Canvas' height in pixels.
0. `color {String / Array / Object}`: **Optional** Color of the canvas. See
   [colors specification](#colors-specification).
0. `inter {String}`: **Optional** interpolation method. Defaults to `"lanczos"`.
   Possible values:
   - `"nearest-neighbor"`
   - `"moving-average"`
   - `"linear"`
   - `"grid"`
   - `"cubic"`
   - `"lanczos"`
0. `callback {Function(err, image)}`

#### Cover

Cover a canvas with the image. The image will be resized to the smallest
possible size such that both its dimensions are bigger than the canvas's
dimensions. Margins of the image exceeding the canvas will be discarded.

`image.cover(width, height, inter, callback)`

0. `width {Integer}`: Canvas' width in pixels.
0. `height {Integer}`: Canvas' height in pixels.
0. `inter {String}`: **Optional** interpolation method. Defaults to `"lanczos"`.
   Possible values:
   - `"nearest-neighbor"`
   - `"moving-average"`
   - `"linear"`
   - `"grid"`
   - `"cubic"`
   - `"lanczos"`
0. `callback {Function(err, image)}`

#### Rotate

`image.rotate(degs, color, callback)`

0. `degs {Float}`: Clockwise rotation degrees.
0. `color {String / Array / Object}`: **Optional** Color of the canvas. See
   [colors specification](#colors-specification).
0. `callback {Function(err, image)}`

#### Crop

#### Crop with rectangle coordinates

`image.crop(left, top, right, bottom, callback)`

0. `left, top, right, bottom {Integer}`: Coordinates of the crop rectangle.
0. `callback {Function(err, image)}`

#### Crop a rectangle from center

`image.crop(width, height, callback)`

0. `width, height {Integer}`: Width and height of the rectangle to crop from the
   center of the image.
0. `callback {Function(err, image)}`

#### Blur

Gaussian blur.

`image.blur(sigma, callback)`

0. `sigma {Float>=0}`: Standard deviation of the Gaussian filter.
0. `callback {Function(err, image)}`

#### Sharpen

Inverse diffusion shapren.

`image.sharpen(amplitude, callback)`

0. `amplitude {Float}`: Sharpening amplitude.
0. `callback {Function(err, image)}`

#### Mirror

Mirror an image along the 'x' axis, 'y' axis or both.

`image.mirror(axes, callback)`

0. `axes {String}`: `'x'`, `'y'` or `'xy'` (case sensitive).
0. `callback {Function(err, image)}`

#### Flip

Alias of [`mirror`](#mirror).

#### Border

Add a colored border to the image.

`image.border(width, color, callback)`

0. `width {Integer}`: Border width in pixels.
0. `color {String / Array / Object}`: **Optional** Color of the border. See
   [colors specification](#colors-specification).
0. `callback {Function(err, image)}`

#### Pad

Pad image edges with colored pixels.

`image.pad(left, top, right, bottom, color, callback)`

0. `left, top, right, bottom {Integer}`: Number of pixels to add to each edge.
0. `color {String / Array / Object}`: **Optional** Color of the padding. See
   [colors specification](#colors-specification).
0. `callback {Function(err, image)}`

#### Saturate

Adjust image saturation.

`image.saturate(delta, callback)`

0. `delta {Float}`: By how much to increase / decrease the saturation.
0. `callback {Function(err, image)}`

**Examples**:

0. `image.saturate(0, ...)` will have no effect on the image.
0. `image.saturate(0.5, ...)` will increase the saturation by 50%.
0. `image.saturate(-1, ...)` will decrease the saturation by 100%, effectively
   desaturating the image.

#### Lighten

Adjust image lightness.

`image.lighten(delta, callback)`

0. `delta {Float}`: By how much to increase / decrease the lightness.
0. `callback {Function(err, image)}`

**Examples**:

0. `image.lighten(0, ...)` will have no effect on the image.
0. `image.lighten(0.5, ...)` will increase the lightness by 50%.
0. `image.lighten(-1, ...)` will decrease the lightness by 100%, effectively
   making the image black.

#### Darken

Adjust image lightness.

`image.darken(delta, callback)`

Equivalent to `image.lighten(-delta, callback)`.

#### Hue

Adjust image hue.

`image.hue(shift, callback)`

0. `shift {Float}`: By how many degrees to shift each pixel's hue.
0. `callback {Function(err, image)}`

**Examples**:

0. `image.lighten(0, ...)` will have no effect on the image.
0. `image.lighten(100, ...)` will shift pixels' hue by 100 degrees.

**Note:** The hue is shifted in a circular manner in the range [0,360] for each
pixel individually.

#### Fade

Adjust image transperancy.

`image.fade(delta, callback)`

0. `delta {Float}`: By how much to increase / decrease the transperancy.
0. `callback {Function(err, image)}`

**Note:** The transparency is adjusted independently for each pixel.

**Examples**:

0. `image.fade(0, ...)` will have no effect on the image.
0. `image.fade(0.5, ...)` will increase the transparency by 50%.
0. `image.fade(1, ...)` will make the image completely transparent.

#### Opacify

Make image completely opaque.

`image.opacify(callback)`

0. `callback {Function(err, image)}`

#### Paste

Paste an image on top of this image.

`image.paste(left, top, img, callback)`

0. `left, top {Integer}`: Coordinates of the top-left corner of the pasted
   image.
0. `img {Image object}`: The image to paste.
0. `callback {Function(err, image)}`

**Notes:**

0. If the pasted image exceeds the bounds of the base image, an exception
   is thrown.
0. `img` is pasted in the state it was at the time `image.paste( ... )` was
   called, eventhough `callback` is called asynchronously.
0. For transparent images, alpha blending is done according to the equations
   described [here](http://en.wikipedia.org/wiki/Alpha_compositing#Alpha_blending).
0. Extra caution is required when using this method in batch mode, as the images
   may change by the time this operation is called.

#### Set Pixel

Set the color of a pixel.

`image.setPixel(left, top, color, callback)`

0. `left, top {Integer}`: Coordinates of the pixel from the left-top corner of
   the image.
0. `color {String / Array / Object}`: Color of the pixel to set.
   See [colors specification](#colors-specification).
0. `callback {Function(err, image)}`

**Notes:**

0. If the coordinates exceed the bounds of the image, an exception is thrown.
0. Extra caution is required when using this method in batch mode, as the
  dimensions of the image may change by the time this operation is called.

#### Set metadata

Set the metadata in an image. This is currently only supported for PNG files.
Sets a tEXt chunk with the key `lwip_data` and comment as the given string. If
called with a `null` parameter, removes existing metadata from the image,
if present.

`image.setMetadata(metadata)`

0. `metadata {String}`: a string of arbitrary length, or null.

CHANGELOG.md
## [3.0.17](https://github.com/randytarampi/lwip/compare/v3.0.16...v3.0.17) (2020-11-22)


### Bug Fixes

* **zlib:** Work around `error: implicit declaration of function 'lseek' is invalid in C99`. Properly. ([da3d750](https://github.com/randytarampi/lwip/commit/da3d750312517e1e046d6d8f62ef7afc16601bec))

## [3.0.16](https://github.com/randytarampi/lwip/compare/v3.0.15...v3.0.16) (2020-11-22)


### Bug Fixes

* **zlib:** Work around `error: implicit declaration of function 'lseek' is invalid in C99`. ([f60acb8](https://github.com/randytarampi/lwip/commit/f60acb8bec26fcdcba2efad430f0c30efef0f02d))

## [3.0.15](https://github.com/randytarampi/lwip/compare/v3.0.14...v3.0.15) (2020-10-31)


### Bug Fixes

* upgrade eslint from 7.11.0 to 7.12.0 ([222b50c](https://github.com/randytarampi/lwip/commit/222b50c748888f03d9b2a26851eacf9afeed3e4c))
* upgrade run-script-os from 1.1.1 to 1.1.3 ([fe7287b](https://github.com/randytarampi/lwip/commit/fe7287b79fa5fffddb5400b537e218fc3cb7dae8))
* upgrade semantic-release from 17.2.1 to 17.2.2 ([8ba2335](https://github.com/randytarampi/lwip/commit/8ba233584092744bd590dfbfaa1bbf31cc10fa78))
* upgrade snyk from 1.416.0 to 1.419.0 ([ebd40c5](https://github.com/randytarampi/lwip/commit/ebd40c59138011134e4e9962bd97cdbd74194e8d))

## [3.0.14](https://github.com/randytarampi/lwip/compare/v3.0.13...v3.0.14) (2020-10-21)


### Bug Fixes

* upgrade snyk from 1.415.0 to 1.416.0 ([c48a917](https://github.com/randytarampi/lwip/commit/c48a917da029e509e0f7aae93aac20360752a12b))

## [3.0.13](https://github.com/randytarampi/lwip/compare/v3.0.12...v3.0.13) (2020-10-20)


### Bug Fixes

* upgrade snyk from 1.414.1 to 1.415.0 ([296608b](https://github.com/randytarampi/lwip/commit/296608b1d01b54dd4dd82b0823563b9f26ec0fa1))

## [3.0.12](https://github.com/randytarampi/lwip/compare/v3.0.11...v3.0.12) (2020-10-18)


### Bug Fixes

* package.json & package-lock.json to reduce vulnerabilities ([bd42ca9](https://github.com/randytarampi/lwip/commit/bd42ca9dbf622a773e6191aa15a2d21b070df126))
* upgrade eslint from 7.10.0 to 7.11.0 ([8f72cff](https://github.com/randytarampi/lwip/commit/8f72cffa0c6a4127ea0450297d1c8b18cbfaa6c0))
* upgrade mocha from 8.1.3 to 8.2.0 ([9fbeaf2](https://github.com/randytarampi/lwip/commit/9fbeaf27131eb3b967afe5552259979c413e6506))
* upgrade nan from 2.14.1 to 2.14.2 ([9b8adc2](https://github.com/randytarampi/lwip/commit/9b8adc21234d38daac8fa156e7e6605be4ddc913))
* upgrade semantic-release from 17.1.2 to 17.2.1 ([f812b98](https://github.com/randytarampi/lwip/commit/f812b981f39b0bd8ebf5de2244cf0c9c0444fddb))
* upgrade snyk from 1.405.1 to 1.406.0 ([74324ce](https://github.com/randytarampi/lwip/commit/74324ce37a445b56a09c99843d3dc311f4b54b94))
* upgrade snyk from 1.406.0 to 1.414.1 ([8b54df3](https://github.com/randytarampi/lwip/commit/8b54df37e384d8a9c5b1082635615e170379d894))

## [3.0.11](https://github.com/randytarampi/lwip/compare/v3.0.10...v3.0.11) (2020-10-03)


### Bug Fixes

* upgrade snyk from 1.402.0 to 1.405.1 ([5d3ebd3](https://github.com/randytarampi/lwip/commit/5d3ebd391b2a476b4f26746888325ec9b2ae0192))

## [3.0.10](https://github.com/randytarampi/lwip/compare/v3.0.9...v3.0.10) (2020-10-01)


### Bug Fixes

* upgrade eslint-plugin-import from 2.22.0 to 2.22.1 ([496c3b2](https://github.com/randytarampi/lwip/commit/496c3b29dece950a152c5f1b7b0e1aa55aab1273))
* upgrade snyk from 1.400.0 to 1.402.0 ([601fbe0](https://github.com/randytarampi/lwip/commit/601fbe088020a103e804dbf20d1a6555e8217691))

## [3.0.9](https://github.com/randytarampi/lwip/compare/v3.0.8...v3.0.9) (2020-09-29)


### Bug Fixes

* upgrade eslint from 7.9.0 to 7.10.0 ([1a908f9](https://github.com/randytarampi/lwip/commit/1a908f9241a5f45221211b13fdb058c2e3c5d3bb))
* upgrade snyk from 1.399.1 to 1.400.0 ([b70718b](https://github.com/randytarampi/lwip/commit/b70718b98134da9da6926bfd8fac1ca8d2d9d3f5))

## [3.0.8](https://github.com/randytarampi/lwip/compare/v3.0.7...v3.0.8) (2020-09-24)


### Bug Fixes

* upgrade snyk from 1.399.0 to 1.399.1 ([834c9d8](https://github.com/randytarampi/lwip/commit/834c9d8973b43aa5fdd3426f7d6597d074730133))

## [3.0.7](https://github.com/randytarampi/lwip/compare/v3.0.6...v3.0.7) (2020-09-23)


### Bug Fixes

* upgrade snyk from 1.398.1 to 1.399.0 ([6205e37](https://github.com/randytarampi/lwip/commit/6205e373dba8b4d6863b01db584956fe4ea637c4))

## [3.0.6](https://github.com/randytarampi/lwip/compare/v3.0.5...v3.0.6) (2020-09-22)


### Bug Fixes

* upgrade semantic-release from 17.1.1 to 17.1.2 ([078bdf2](https://github.com/randytarampi/lwip/commit/078bdf2a1b604e6a000905970d2bef18deadbb45))

## [3.0.5](https://github.com/randytarampi/lwip/compare/v3.0.4...v3.0.5) (2020-09-18)


### Bug Fixes

* upgrade snyk from 1.395.0 to 1.398.1 ([3739786](https://github.com/randytarampi/lwip/commit/3739786b16df86d08faaf224f20e07763d2ea079))

## [3.0.4](https://github.com/randytarampi/lwip/compare/v3.0.3...v3.0.4) (2020-09-17)


### Bug Fixes

* upgrade snyk from 1.393.0 to 1.395.0 ([30ec568](https://github.com/randytarampi/lwip/commit/30ec56875c9872f65a2b7ede0724435e54002450))

## [3.0.3](https://github.com/randytarampi/lwip/compare/v3.0.2...v3.0.3) (2020-09-14)


### Bug Fixes

* upgrade eslint from 7.8.1 to 7.9.0 ([0f8664b](https://github.com/randytarampi/lwip/commit/0f8664b73e2d8341b663ba044f677b1338226215))
* upgrade snyk from 1.387.0 to 1.393.0 ([5e52057](https://github.com/randytarampi/lwip/commit/5e52057eb723973f3955e180706a5a9cadcc2e6a))

## [3.0.2](https://github.com/randytarampi/lwip/compare/v3.0.1...v3.0.2) (2020-09-13)


### Bug Fixes

* upgrade eslint from 7.7.0 to 7.8.1 ([8c9d1dd](https://github.com/randytarampi/lwip/commit/8c9d1ddba3055bc06dcc169e5dd69a77311de9de))
* upgrade snyk from 1.386.0 to 1.387.0 ([19de674](https://github.com/randytarampi/lwip/commit/19de674f18c16f68f47c4fd401d6c5cbcc9d13d5))

## [3.0.1](https://github.com/randytarampi/lwip/compare/v3.0.0...v3.0.1) (2020-08-29)


### Bug Fixes


lib/defs.js
'use strict';

const defaults = exports.defaults = {
    DEF_INTERPOLATION: 'lanczos',
    DEF_ROTATE_COLOR: 'gray',
    DEF_BORDER_COLOR: 'gray',
    DEF_PAD_COLOR: 'gray',
    DEF_COLOR_ALPHA: 100,
    DEF_JPEG_QUALITY: 100,
    PNG_DEF_COMPRESSION: 'fast',
    PNG_DEF_INTERLACED: false,
    PNG_DEF_TRANSPARENT: 'auto',
    GIF_DEF_COLORS: 256,
    GIF_DEF_INTERLACED: false,
    GIF_DEF_TRANSPARENT: 'auto',
    GIF_DEF_THRESHOLD: 50,
    DEF_CREATE_COLOR: [0, 0, 0, 0],
    DEF_CONTAIN_COLOR: [0, 0, 0, 0]
};

exports.interpolations = {
    'nearest-neighbor': 1,
    'moving-average': 2,
    'linear': 3,
    'grid': 4,
    'cubic': 5,
    'lanczos': 6
};

exports.colors = {
    'black': {
        r: 0,
        g: 0,
        b: 0,
        a: 100
    },
    'white': {
        r: 255,
        g: 255,
        b: 255,
        a: 100
    },
    'red': {
        r: 255,
        g: 0,
        b: 0,
        a: 100
    },
    'blue': {
        r: 0,
        g: 0,
        b: 255,
        a: 100
    },
    'green': {
        r: 0,
        g: 255,
        b: 0,
        a: 100
    },
    'cyan': {
        r: 0,
        g: 255,
        b: 255,
        a: 100
    },
    'yellow': {
        r: 255,
        g: 255,
        b: 0,
        a: 100
    },
    'gray': {
        r: 128,
        g: 128,
        b: 128,
        a: 100
    },
    'magenta': {
        r: 255,
        g: 0,
        b: 255,
        a: 100
    }
};

exports.args = {
    open: [{
        name: 'source',
        type: '*'
    }, {
        name: 'type',
        types: ['string', 'raw-buffer-properties'],
        optional: true
    }, {
        name: 'callback',
        type: 'function'
    }],
    create: [{
        name: 'width',
        type: 'p-number'
    }, {
        name: 'height',
        type: 'p-number'
    }, {
        name: 'color',
        type: 'color',
        optional: true,
        default: defaults.DEF_CREATE_COLOR
    }, {
        name: 'callback',
        type: 'function'
    }],
    scale: [{
        name: 'wRation',
        type: 'p-number'
    }, {
        name: 'hRatio',
        type: 'p-number',
        optional: true
    }, {
        name: 'interpolation',
        type: 'interpolation',
        optional: true,
        default: defaults.DEF_INTERPOLATION
    }, {
        name: 'callback',
        type: 'function'
    }],
    resize: [{
        name: 'width',
        type: 'p-number'
    }, {
        name: 'height',
        type: 'p-number',
        optional: true
    }, {
        name: 'interpolation',
        type: 'interpolation',
        optional: true,
        default: defaults.DEF_INTERPOLATION
    }, {
        name: 'callback',
        type: 'function'
    }],
    rotate: [{
        name: 'degs',
        type: 'number'
    }, {
        name: 'color',
        type: 'color',
        optional: true,
        default: defaults.DEF_ROTATE_COLOR
    }, {
        name: 'callback',
        type: 'function'
    }],
    blur: [{
        name: 'sigma',
        type: 'nn-number'
    }, {
        name: 'callback',
        type: 'function'
    }],
    hslaAdjust: [{
        name: 'hs',
        type: 'number'
    }, {
        name: 'sd',
        type: 'number'
    }, {
        name: 'ld',
        type: 'number'
    }, {
        name: 'ad',
        type: 'number'
    }, {
        name: 'callback',
        type: 'function'
    }],
    saturate: [{
        name: 'delta',
        type: 'number'
    }, {
        name: 'callback',
        type: 'function'
    }],
    lighten: [{
        name: 'delta',
        type: 'number'
    }, {
        name: 'callback',
        type: 'function'
    }],
    darken: [{
        name: 'delta',
        type: 'number'
    }, {
        name: 'callback',
        type: 'function'
    }],
    fade: [{
        name: 'delta',
        type: 'number'
    }, {
        name: 'callback',
        type: 'function'
    }],
    opacify: [{
        name: 'callback',
        type: 'function'
    }],
    hue: [{
        name: 'shift',
        type: 'number'
    }, {
        name: 'callback',
        type: 'function'
    }],
    crop: [{
        name: 'left',
        type: 'nn-number'
    }, {
        name: 'top',
        type: 'nn-number'
    }, {
        name: 'right',
        type: 'nn-number',
        optional: true
    }, {
        name: 'bottom',
        type: 'nn-number',
        optional: true
    }, {
        name: 'callback',
        type: 'function'
    }],
    mirror: [{
        name: 'axes',
        type: 'axes'
    }, {
        name: 'callback',
        type: 'function'
    }],
    exec: [{
        name: 'callback',
        type: 'function'
    }],
    toBuffer: [{
        name: 'type',
        type: 'string'
    }, {
        name: 'params',
        type: 'hash',
        optional: true,
        default: {}
    }, {
        name: 'callback',
        type: 'function'
    }],
    getPixel: [{
        name: 'left',
        type: 'nn-int'
    }, {
        name: 'top',
        type: 'nn-int'
    }],
    border: [{
        name: 'width',
        type: 'nn-number'
    }, {
        name: 'color',
        type: 'color',
        optional: true,
        default: defaults.DEF_BORDER_COLOR
    }, {
        name: 'callback',
        type: 'function'
    }],
    pad: [{
        name: 'left',
        type: 'nn-number'
    }, {
        name: 'top',
        type: 'nn-number'
    }, {
        name: 'right',
        type: 'nn-number'
    }, {
        name: 'bottom',
        type: 'nn-number'
    }, {
        name: 'color',
        type: 'color',
        optional: true,
        default: defaults.DEF_PAD_COLOR
    }, {
        name: 'callback',
        type: 'function'
    }],
    sharpen: [{
        name: 'amplitude',
        type: 'number'
    }, {
        name: 'callback',
        type: 'function'
    }],
    clone: [{
        name: 'callback',
        type: 'function'
    }],
    paste: [{
        name: 'left',
        type: 'nn-number'
    }, {
        name: 'top',
        type: 'nn-number'
    }, {
        name: 'image',
        type: 'image'
    }, {
        name: 'callback',
        type: 'function'
    }],
    extract: [{
        name: 'left',
        type: 'nn-number'
    }, {
        name: 'top',
        type: 'nn-number'
    }, {
        name: 'right',
        type: 'nn-number'
    }, {
        name: 'bottom',
        type: 'nn-number'
    }, {
        name: 'callback',
        type: 'function'
    }],
    writeFile: [{
        name: 'path',
        type: 'string'
    }, {
        name: 'type',
        type: 'string',
        optional: true
    }, {
        name: 'params',
        type: 'hash',
        optional: true,
        default: {}
    }, {
        name: 'callback',
        type: 'function'
    }],
    setPixel: [{
        name: 'left',
        type: 'nn-int'
    }, {
        name: 'top',
        type: 'nn-int'
    }, {
        name: 'color',
        type: 'color'
    }, {
        name: 'callback',
        type: 'function'
    }],
    contain: [{
        name: 'width',
        type: 'p-number'
    }, {
        name: 'height',
        type: 'p-number'
    }, {
        name: 'color',
        type: 'color',
        optional: true,
        default: defaults.DEF_CONTAIN_COLOR
    }, {
        name: 'interpolation',
        type: 'interpolation',
        optional: true,
        default: defaults.DEF_INTERPOLATION
    }, {
        name: 'callback',
        type: 'function'
    }],
    cover: [{
        name: 'width',
        type: 'p-number'
    }, {
        name: 'height',
        type: 'p-number'
    }, {
        name: 'interpolation',
        type: 'interpolation',
        optional: true,
        default: defaults.DEF_INTERPOLATION
    }, {
        name: 'callback',
        type: 'function'
    }],
};
